pkg/safetime/safetime.go

```diff
@@ -0,0 +1,38 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safetime
+
+import (
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// TimeSinceSafe returns the duration since t. If the duration is negative,
+// returns false to indicate the fact.
+//
+// Used to workaround a malfunctioning monotonic clock.
+func TimeSinceSafe(t time.Time, logger *log.Entry) (bool, time.Duration) {
+	n := time.Now()
+	d := n.Sub(t)
+
+	if d < 0 {
+		logger.WithField("then", t).WithField("now", n).WithField("duration", d).
```

Comment:
The logrus objects also have a `.WithFields` function that takes a `logrus.Fields`type (and that is basically a `map[string]interface{}`. 
Relatedly, for messages at Info and above, which is to say that they are seen during normal operation, we try to use consistent field names from the logfields package. In this case, we already have `startTime` and `duration`  as field names (https://github.com/cilium/cilium/blob/master/pkg/logging/logfields/logfields.go#L108-L112). We can add an `endTime` to round it out and use all three here.